compiler/src/iree/compiler/Dialect/HAL/Transforms/SerializeExecutables.cpp

```diff
@@ -162,13 +173,15 @@ class SerializeExecutablesPass
  private:
   std::string dumpIntermediatesPath;
   std::string dumpBinariesPath;
+  std::string dumpedSourcesPath;
```

Comment:
This spidering isn't great - I think you're leaning a bit too much into reusing things :)

ExecutableExportOp gaining an optional LocationAttr or something would let us flexibly decide how to annotate the exports (could come all the way from the user) and be stored in the IR where it doesn't need this plumbing. When we want to change to using a full program source listing instead of doing it with individual files we'll need that anyway.

With your current behavior this would mean the DumpExecutableSources pass would annotate the exports with the attribute as it runs and then the library serialization code would fetch that from the export it was processing (and other backends could do the same).

For now you could just support FileLineColLoc and get the file/line from it and ignore others - in the future we could export something like we do for the vm debug databases that encodes the full source information if we wanted (line/col ranges/etc).

Reply:
Is there a specific way that you suggest attaching a `FileLineColLoc` to `HAL_ExecutableExportOp`? I've tried a few different ways to use what's in https://github.com/llvm/llvm-project/blob/26dd42705c2af0b8f6e5d6cdb32c9bd5ed9524eb/mlir/include/mlir/IR/BuiltinLocationAttributes.td#L65-L101 at https://github.com/iree-org/iree/blob/a0943ed99ea31723b839a2234a3070f46d5f2bab/compiler/src/iree/compiler/Dialect/HAL/IR/HALOps.td#L1647-L1654, but so far all I have working is `OptionalAttr<StrAttr>:$source_location`. Once I'm in C++ I can use `FileLineColLoc::get(&getContext(), filePath, 0, 0);` to get a `FileLineColLoc`, but then I'd somehow need to cast that to/from a string if I'm stuck with `StrAttr`

Reply:
Still not finding a good way to add an optional LocationAttr, but I think I can just use the standard LocationAttr that all ops have, together with the snapshot created here: https://github.com/iree-org/iree/blob/36f2451913c025c725f2d7e03935d2e9da9a872c/compiler/src/iree/compiler/Dialect/HAL/Transforms/DumpExecutableSources.cpp#L25-L31

then DumpExecutableSources is just _one_ way to get location data - if you happen to have the python source or something on disk and don't snapshot with that pass maybe that will naturally work too?

Reply:
Yep, that worked, but I'm not fully satisfied with how relative and absolute paths are handled. I had to switch that snapshot from using file names to using the full file paths (without normalization), so relative paths have a chance of making their way through to Tracy.